harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -646,27 +1031,38 @@ def _load(self, load_path: pathlib.Path) -> None:
                     "callback will be initialized from scratch"
                 )
 
-        # Load workload sequencer state.
-        wlsq_path = load_path.joinpath("workload_sequencer.pkl")
-        if self.wlsq is not None and wlsq_path.exists():
-            with wlsq_path.open("rb") as f:
-                self.wlsq.load_state(pickle.load(f))
+        save_path = load_path.joinpath("trial_state.pkl")
+        if save_path.exists():
+            with save_path.open("rb") as f:
+                self._load_state(pickle.load(f))
+
+    def _load_state(self, state: Any) -> None:
+        # Load our state from the checkpoint if we are continuing training after a pause or restart.
+        # If the trial_id doesn't match our current trial id, we're continuing training a previous
+        # trial and should start from a fresh state.
+        if state.get("trial_id") != self.trial_id:
+            self.state = pytorch.TrialState(trial_id=self.trial_id)
+            return
+
+        self.state = pytorch.TrialState(**state)
+        assert self.state
+
+        # Detect the case where the final validation we made was against this exact checkpoint.  In
+        # that case, the master will know about the validation, but it would not appear in the
+        # checkpoint state.  If the validation was before the last checkpoint, the checkpoint state
+        # is already correct, while any validations after the last checkpoint aren't valid anymore
+        # and can be safely ignored.
+        if self.state.batches_trained == self.val_from_previous_run:
+            self.state.last_val = self.state.batches_trained
 
     def _save(self, path: pathlib.Path) -> None:
-        if self.context.distributed.local_rank == 0:
-            path.mkdir(parents=True, exist_ok=True)
-        _ = self.context.distributed.gather_local(None)  # sync
+        path.mkdir(parents=True, exist_ok=True)
 
         if self.is_chief:
             # We assume these stateful objects should be the same across slots and only have
             # the chief save them.
-            util.write_user_code(path, self.env.on_cluster)
-
-            if self.wlsq is not None:
```

Comment:
>             # We assume these stateful objects should be the same across slots and only have
>             # the chief save them.

not a big deal, but previously we were saving the training state only on the chief (now TrialState)

Reply:
DeepSpeedTrial previously called `_save()` on every rank, so each gpu had it's own state. The only thing done once was/is saving user code.

Reply:
but the saving of the training state is also gated by the chief worker?

```
        if self.is_chief:
            # We assume these stateful objects should be the same across slots and only have
            # the chief save them.
            util.write_user_code(path, self.env.on_cluster)

            if self.wlsq is not None:
                with path.joinpath("workload_sequencer.pkl").open("wb") as f:
                    pickle.dump(self.wlsq.get_state(), f)

```

Reply:
added state to the top `self.is_chief` portion of this method